src/Google.Storage.V1/CopyObjectOptions.cs

```diff
@@ -0,0 +1,162 @@
+
+        /// <summary>
+        /// Precondition for copying: the object is only copied if the existing destination object's
+        /// meta-generation does not match the given value.
+        /// </summary>
+        public long? IfMetagenerationNotMatch { get; set; }
+
+        /// <summary>
+        /// Precondition for copying: the object is only copied if the source object's
+        /// generation matches the given value.
+        /// </summary>
+        public long? IfSourceGenerationMatch { get; set; }
+
+        /// <summary>
+        /// Precondition for copying: the object is only copied if the source object's
+        /// generation does not match the given value.
+        /// </summary>
+        public long? IfSourceGenerationNotMatch { get; set; }
+
+        /// <summary>
+        /// Precondition for copying: the object is only copied if the source object's
+        /// meta-generation matches the given value.
+        /// </summary>
+        public long? IfSourceMetagenerationMatch { get; set; }
+
+        /// <summary>
+        /// Precondition for copying: the object is only copied if the source object's
+        /// meta-generation does not match the given value.
+        /// </summary>
+        public long? IfSourceMetagenerationNotMatch { get; set; }
+
+        internal void ModifyRequest(RewriteRequest request)
+        {
+            // Note the use of ArgumentException here, as this will basically be the result of invalid
+            // options being passed to a public method.
+            if (IfGenerationMatch != null && IfGenerationNotMatch != null)
+            {
+                throw new ArgumentException($"Cannot specify {nameof(IfGenerationMatch)} and {nameof(IfGenerationNotMatch)} in the same options", "options");
+            }
+            if (IfMetagenerationMatch != null && IfMetagenerationNotMatch != null)
+            {
+                throw new ArgumentException($"Cannot specify {nameof(IfMetagenerationMatch)} and {nameof(IfMetagenerationNotMatch)} in the same options", "options");
+            }
+            if (IfSourceGenerationMatch != null && IfSourceGenerationNotMatch != null)
+            {
+                throw new ArgumentException($"Cannot specify {nameof(IfSourceGenerationMatch)} and {nameof(IfSourceGenerationNotMatch)} in the same options", "options");
+            }
+            if (IfSourceMetagenerationMatch != null && IfSourceMetagenerationNotMatch != null)
+            {
+                throw new ArgumentException($"Cannot specify {nameof(IfSourceMetagenerationMatch)} and {nameof(IfSourceMetagenerationNotMatch)} in the same options", "options");
+            }
+
+            if (DestinationPredefinedAcl != null)
+            {
+                request.DestinationPredefinedAcl =
+                    Preconditions.CheckEnumValue((DestinationPredefinedAclEnum)DestinationPredefinedAcl, nameof(DestinationPredefinedAcl));
+            }
+            if (SourceGeneration != null)
```

Comment:
I'm not clear on the expected behaviour here.
request.SourceGeneration is nullable, so is the 'if' required? Or is it only wanting to override the values in 'request' if this CopyObjectOptions.SourceGeneration value is not null?